OpTestSystem: Connection enhancements

debmc opened the pull request:
Enhancements included in this patch:

Connection handling tolerant of glitches.

Refactor modules to single source shared utilities.

Automatic login and prompt setup for consoles.

Signed-off-by: Deb McLemore <debmc@linux.vnet.ibm.com>

pridhiviraj commented:
```

 *** WaitForIt CURRENT STATE "02" TARGET STATE "06"
 *** WaitForIt working on transition
 *** Current loop iteration "149"             - Reconnect attempts "02" - loop_max "150"
 *** WaitForIt timeout interval "04" seconds - Stale buffer check every "02" times
 *** WaitForIt variables "['/ #', 'Petitboot', 'System shutting down with error status', 'login: ', 'mon> ', 'shutdown requested', 'x=exit']"
 *** WaitForIt Refresh="0" Buffer Kicker="0" - Kill Cord="150"


 *** WaitForIt CURRENT STATE "02" TARGET STATE "06"
 *** WaitForIt working on transition
 *** Current loop iteration "150"             - Reconnect attempts "02" - loop_max "150"
 *** WaitForIt timeout interval "04" seconds - Stale buffer check every "02" times
 *** WaitForIt variables "['/ #', 'Petitboot', 'System shutting down with error status', 'login: ', 'mon> ', 'shutdown requested', 'x=exit']"
 *** WaitForIt Refresh="0" Buffer Kicker="0" - Kill Cord="150"


#IPMI SOL CONNECT
~.[SOL Session operational.  Use ~? for help]
OpTestSystem has reached the limit on re-IPL'ing to try to recover, we will be stopping
```
On a P8 habanero system, it is not able to IPL the system, and also the above print messages of WaitForIt should be in debug level not in print level.

pridhiviraj commented:
State 2 to 4 transition is entered into a loop forever.
```
[SOL Session operational.  Use ~? for help]

 *** WaitForIt CURRENT STATE "02" TARGET STATE "04"
 *** WaitForIt working on transition
 *** Current loop iteration "89"             - Reconnect attempts "43" - loop_max "100"
 *** WaitForIt timeout interval "04" seconds - Stale buffer check every "02" times
 *** WaitForIt variables "['/ #', 'Petitboot', 'System shutting down with error status', 'login: ', 'mon> ', 'shutdown requested', 'x=exit']"
 *** WaitForIt Refresh="0" Buffer Kicker="0" - Kill Cord="150"


 *** WaitForIt CURRENT STATE "02" TARGET STATE "04"
 *** WaitForIt working on transition
 *** Current loop iteration "90"             - Reconnect attempts "43" - loop_max "100"
 *** WaitForIt timeout interval "04" seconds - Stale buffer check every "02" times
 *** WaitForIt variables "['/ #', 'Petitboot', 'System shutting down with error status', 'login: ', 'mon> ', 'shutdown requested', 'x=exit']"
 *** WaitForIt Refresh="0" Buffer Kicker="0" - Kill Cord="150"
```

debmc commented:
> State 2 to 4 transition is entered into a loop forever.

If you can provide a system I can debug, is there anything special about system that would be different that you know of ?

debmc commented:
> WaitForIt should be in debug level not in print level.

I thought the output was helpful to know what was happening instead of watching nothing for a long time, it also provides information on what state or stage the machine is in.

debmc commented:
> On a P8 habanero system, it is not able to IPL the system

Fixed, I missed catching an EOF, I posted a new patch.

debmc commented:
> Two thoughts so far. I'm going to try and do some more testing with it tomorrow.

Added the export LANG=C and failed the OpalGard case and fixed OpTestHost to run the fake guard.

pridhiviraj commented on common/OpTestIPMI.py:
```diff
@@ -160,64 +161,118 @@ def set_system_to_UNKNOWN_BAD(system):
-            self.sol.terminate()
-            self.state = IPMIConsoleState.DISCONNECTED
+          rc_child = self.close()
+          self.state = IPMIConsoleState.DISCONNECTED
```
self.sol.close() already takes care of setting the console state to DISCONNECTED, so in the below three cases you can remove state set to DISCONNECTED line.

debmc replied:
> remove state set to DISCONNECTED line

Fixed 3 cases.

pridhiviraj commented on common/OpTestIPMI.py:
```diff
@@ -160,64 +161,118 @@ def set_system_to_UNKNOWN_BAD(system):
+          return solChild
+        if rc == 1:
+          self.sol.close()
+          self.state = IPMIConsoleState.DISCONNECTED
```
Same here as well.

pridhiviraj commented on common/OpTestIPMI.py:
```diff
@@ -160,64 +161,118 @@ def set_system_to_UNKNOWN_BAD(system):
+          raise CommandFailed('sol activate', "IPMI: pexpect.TIMEOUT while trying to establish connection", -1)
+        if rc == 2:
+          self.sol.close()
+          self.state = IPMIConsoleState.DISCONNECTED
```
same here as well.

pridhiviraj commented on common/OpTestSystem.py:
```diff
@@ -222,91 +385,182 @@ def detect_target(self, target_state, reboot):
+                      " *** WaitForIt variables \"{}\"\n"
+                      " *** WaitForIt Refresh=\"{}\" Buffer Kicker=\"{}\" - Kill Cord=\"{:02}\"\n".format(self.state, self.target_state,
+                      x, reconnect_count, kwargs['loop_max'], kwargs['timeout'], kwargs['threshold'],
+                      sorted(kwargs['expect_dict'].keys()), kwargs['refresh'], kwargs['buffer_kicker'], self.kill_cord))
```
Atleast comment this statement for now, until your logging infrastructure PR merges.

debmc replied:
> comment this statement for now

Fixed

pridhiviraj commented on common/Exceptions.py:
```diff
@@ -201,3 +201,97 @@ class HostbootShutdown(Exception):
+           ' review the following for more details\nExpect Before Buffer=\"{}\"\nExpect After Buffer=\"{}\"'
+           '\nMessage=\"{}\"'.format(self.kwargs['before'], self.kwargs['after'], self.kwargs['msg']))
+
+class UnknownSystem(Exception):
```
This should be UnknownStateTransition? As that is matching as per the error conditions.

pridhiviraj commented on common/OpTestSystem.py:
```diff
@@ -222,91 +385,182 @@ def detect_target(self, target_state, reboot):
 
-    def wait_for_it(self, my_string=None, refresh=1, buffer_kicker=1, loop_max=50, threshold=1, timeout=10):
-        console = self.console.get_console()
+    def wait_for_it(self, **kwargs):
```
I re-tested this latest code on a P9 system, and i see it is failing to reach petitboot state(3) as it is missing the string. This we need to really improve or atleast for now block the re-connect attempts as long as the console is stable we no need to worry for re-connecting. You can keep this new variant logic as optional, and use old method as a stable variant and let callers use this stable variant instead of optional one.

debmc replied:
> So it is in a loop forever even though system actually reached petitboot.

Yes, I found P9 system console output that gets to pexpect buffer can be very slow and so anytime it can miss catching anything, so I think the last_try helps catch and also that is what the logic of catching login is unexpected when trying to reach Petitboot and vice-versa if catching Petitboot if trying to go to login.  I think its the best case and it can never compensate for not getting data (either from slow console or from drops of connection like IPMI SOL).

I'll revisit and see if we can tune the non-IPMI cases a little better.

debmc replied:
> failing to reach petitboot state(3)

I also found on OpenBMC that if I tried manually through the port 2200 console it would not properly stop at the Petitboot if bootdev was set to stop, it would always go to OS.  Maybe check your system and see if this is happening.  If the system was set to stop at Petitboot then the buffer being slow is sync'd up and wait_for_it catches.

pridhiviraj replied:
My system is a boston system, which has all this functionality working, but with this new logic it may be failed to find petitboot string.
```
[   90.407191272,5] INIT: 64-bit LE kernel discovered
[   90.409329218,5] OCC: All Chip Rdy after 0 ms
[   90.438891161,5] INIT: Starting kernel at 0x20011000, fdt at 0x30a97f80 364786 bytes
[   92.111606245,5] OPAL: Switch to little-endian OS
ipmitool -H $BMCIP -I lanplus -U ADMIN -P ADMIN sol deactivate
#IPMI SOL CONNECT
[SOL Session operational.  Use ~? for help]
~. [terminated ipmitool]
ipmitool -H $BMCIP -I lanplus -U ADMIN -P ADMIN sol deactivate
#IPMI SOL CONNECT
[SOL Session operational.  Use ~? for help]
~. [terminated ipmitool]
ipmitool -H $BMCIP -I lanplus -U ADMIN -P ADMIN sol deactivate
#IPMI SOL CONNECT
[SOL Session operational.  Use ~? for help]
~. [terminated ipmitool]
ipmitool -H $BMCIP -I lanplus -U ADMIN -P ADMIN sol deactivate
#IPMI SOL CONNECT
[SOL Session operational.  Use ~? for help]
~. [terminated ipmitool]
ipmitool -H $BMCIP -I lanplus -U ADMIN -P ADMIN sol deactivate
#IPMI SOL CONNECT
[SOL Session operational.  Use ~? for help]
~. [terminated ipmitool]
ipmitool -H $BMCIP -I lanplus -U ADMIN -P ADMIN sol deactivate
#IPMI SOL CONNECT
[SOL Session operational.  Use ~? for help]
~. [terminated ipmitool]
```

After that i tried checking what is the state of the system.
it is in petitboot shell.
```
[SOL Session operational.  Use ~? for help]

/ # 
```

debmc replied:
> it is in petitboot shell.

If you can give me the system I can debug to check if its something new or if its the same cases I have seen.

debmc commented:
> WaitForIt logic

Tuned the parameters for an optimal case across a variety of platforms.

Running BasicIPL on Boston:

Current REPO:
1131.461s
1130.447s
1131.020s

With Patch:
1167.159s
1168.122s
1168.487s

stewart-ibm commented:
Deb McLemore <notifications@github.com> writes:
>> Connection handling tolerant of glitches.
>
> Just some Observations on IPMI SOL and OPexpect.spawn handling:
>
> When the OPexpect spawn process stops seeing data coming in its buffers, op-test can detect staleness, nothing is changing in the pexpect buffer.  The only way to kick this is to send a buffer kicker, sometimes repeatedly (like hitting the enter key if you use manual means).  Sending sporadic enter keys during the boot cycle when it gets close to Petitboot can disrupt the transition to get to booting the OS (the last_try mechanism is used to pretend like an end-user and hit the enter key, etc, which will kick things free and hopefully the transition is in a good place to perform such actions).
>
> One way to overcome this was to get a completely fresh OPexpect spawn process (which for IPMI deactivates/activates SOL) and then data will automatically start flowing in the OPexpect spawn process buffer without having to kick the buffer with any extraneous sends (but this doesn't always happen).
>
> You can manually use ipmitool commands to power off and then power on
> the box, which also seems to sync things up, I have had SOL itself
> stop responding and lock up manually, very unpredictable, and even
> needing to reboot the BMC, such fun (and even flop to golden side,
> etc).

Yeah, my horrible code to reconnect on SOL disconnect did this, and that
"fixed" a lot of things on AMI BMC based P8 systems. Otherwise we'd just
hang or reconnect forever and never get anywhere.

Mind you, I wonder how much we work around these BMC bugs and just say
"you have to fix your BMC implementation".

> However, sometimes a new OPexpect spawn process seems to get initially
> hung up without anything in the buffer (I read a few things about bugs
> in pexpect with the socket and issues with the buffer pointer locking
> up, etc).  The only way I overcame that issue in other places is to
> sendline("date") which actually produces data in the new incoming
> pexpect buffer which seems to sync things up and works.  I found a
> comment in some of the pexpect code about this being like "playing in
> traffic".

Hrm... I've not seen this myself, at least not when creating a new
pexpect object for each new connection. I wonder if there were any bugs
fixed in pexpect between what we've been using.

-- 
Stewart Smith
OPAL Architect, IBM.

debmc commented:
> creating a new pexpect object

I first thought this issue was timing, a delay between the console data making its way to the pexpect buffer, but even on fresh pexpect process adding long sleeps between sendline("date") methodically to time it, it would not start showing the output of the date command in the buffer at all.  I finally squeezed the issue down to set_PS1 method and check_root method.  Another pattern I observed was that immediately following a pexpect.TIMEOUT (when I would see the pexpect.TIMEOUT in the AFTER output is when it typically froze up and typically always while in Petitboot). The get_sudo method will always timeout in Petitboot and if running as root so I debugged those and could not figure out how to circumvent.

debmc commented:
> "you have to fix your BMC implementation".

+1

debmc commented:
> Sample QEMU

For direct invocation of skiroot with QEMU:

sudo /home/debmc/qemu/ppc64-softmmu/qemu-system-ppc64 -machine powernv -m 4G -nographic -nodefaults -bios /home/debmc/op-build/output/images/skiboot.lid -kernel /home/debmc/op-build/output/images/zImage.epapr -initrd /home/debmc/op-build/output/images/rootfs.cpio -device ipmi-bmc-sim,id=bmc0 -device isa-ipmi-bt,bmc=bmc0,irq=10 -device isa-serial,chardev=s1 -chardev stdio,id=s1,signal=off -nic user,model=virtio-net-pci

Exiting petitboot. Type 'exit' to return.
You may run 'pb-sos' to gather diagnostic data
/ # ip add
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: tunl0@NONE: <NOARP> mtu 1480 qdisc noop qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
3: enp2s1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast qlen 1000
    link/ether 52:54:00:12:34:56 brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global enp2s1
       valid_lft forever preferred_lft forever
/ # poweroff                                       <<<<<<<<<<<  Since signals are off, easy way to exit QEMU
The system is going down NOW!
Sent SIGTERM to all processes
Sent SIGKILL to all processes
Requesting system poweroff
[   27.908580] reboot: Power down
[   51.248468531,5] OPAL: Shutdown request type 0x0...

stewart-ibm commented:
So, at this point I'm pretty happy with this, and would bet that it's better to go forward with what's here than what's in tree - so with two small fixes, merged!
